About Us
================================

InterMine is an open source data warehouse system for the integration and analysis of complex biological data, developed for the last 10 years by the Micklem Lab at the University of Cambridge. InterMine has been used for developing data warehousing solutions for a number of projects, including for storage and analysis of modENCODE data, and as a data mining platform for a number of major model organism databases as part of the InterMOD project. 

InterMine has been developed with the support of the Wellcome Trust [067205], [082598], [090297], as well as support from the National Human Genome Research Institute [R01HG004834]. The Wellcome Trust also recently granted a further 5 years of funding for InterMine development, as well as development of HumanMine, a data warehouse of human genetic, genomic and proteomic data, ensuring continued development of InterMine as a framework. 


The publicly available InterMine instances include: 

* `FlyMine <http://www.flymine.org>`_  - a data warehouse of integrated fruit fly genetic, genomic and proteomic data 
* `YeastMine <http://yeastmine.yeastgenome.org>`_ - an integrated data warehouse of yeast genomic data, developed by SGD
* `RatMine <http://ratmine.mcw.edu/ratmine>`_ - an integrated data warehouse of rat genomic data, developed by RGD
* `MouseMine <http://www.mousemine.org>`_ - an integrated data warehouse of mouse genomic data, developed by MGI
* `TargetMine <https://targetmine.mizuguchilab.org/targetmine>`_ - a data warehouse for candidate gene prioritisation and drug target discovery, developed at NIBIO, Japan
* `Zebrafishmine <http://www.Zebrafishmine.org>`_ - an integrated data warehouse of zebrafish genomic data, developed by ZFIN
* `Thalemine <https://bar.utoronto.ca/thalemine>`_ - a data warehouse for Arabidopsis thaliana Col-0 for the ARAPORT project
* `PhytoMine <https://phytozome.jgi.doe.gov/phytomine>`_ - an integrated data warehouse of over 50 plant genomes from Phytozome.

See the InterMine `registry <http://registry.intermine.org>`_ for the full list of InterMine instances.

More information:

.. toctree::
    :maxdepth: 2

    contact-us


How to cite us
------------------------

If you use the InterMine framework in your research, we would appreciate it if you cite the following publication: 

* InterMine: extensive web services for modern biology. Kalderimis A, Lyne R, Butano D, Contrino S, Lyne M, Heimbach J, Hu F, Smith R, Stěpán R, Sullivan J, Micklem G. Nucleic Acids Res. 2014 Jul; 42 (Web Server issue): W468-72

* InterMine: a flexible data warehouse system for the integration and analysis of heterogeneous biological data. Smith RN, Aleksic J, Butano D, Carr A, Contrino S, Hu F, Lyne M, Lyne R, Kalderimis A, Rutherford K, Stepan R, Sullivan J, Wakeling M, Watkins X, Micklem G. Bioinformatics (2012) 28 (23): 3163-3165.

See the InterMine `zotero <https://www.zotero.org/groups/2117194/intermine/>`_ group for the full list of InterMine publications.

.. _legal:

Legal
------

All InterMine code is freely available under the open source `LGPL <http://www.gnu.org/licenses/lgpl.html>`_  license.

Privacy Policy
------------------

.. toctree::
    :maxdepth: 2

    privacy-policy

.. index:: LGPL, jobs, careers, hiring, vacancies, privacy policy, release notes, cite, about us
